src/include/neorados/RADOS.hpp

```diff
@@ -745,82 +719,43 @@ class RADOS final
-	     std::optional<std::chrono::seconds> timeout,
-	     WatchCB&& cb, CompletionToken&& token) {
-    boost::asio::async_completion<CompletionToken, WatchSig> init(token);
-    watch(o, ioc, timeout, std::move(cb),
-	  WatchComp::create(get_executor(),
-			    std::move(init.completion_handler)));
-    return init.result.get();
-  }
-
-  template<typename CompletionToken>
-  auto watch(const Object& o, std::int64_t pool,
+  auto watch(Object o, IOContext ioc,
 	     std::optional<std::chrono::seconds> timeout,
-	     WatchCB&& cb, CompletionToken&& token,
-	     std::optional<std::string_view> ns = {},
-	     std::optional<std::string_view> key = {}) {
-    boost::asio::async_completion<CompletionToken, WatchSig> init(token);
-    watch(o, pool, timeout, std::move(cb),
-	  WatchComp::create(get_executor(),
-			    std::move(init.completion_handler)),
-	  ns, key);
-    return init.result.get();
+	     WatchCB cb, CompletionToken&& token) {
+    return boost::asio::async_initiate<CompletionToken, WatchSig>(
+      [o = std::move(o), ioc = std::move(ioc), timeout, cb = std::move(cb),
+       this](auto&& handler) mutable {
+	watch(std::move(o), std::move(ioc), timeout, std::move(cb),
+	      WatchComp::create(get_executor(),
+				std::move(handler)));
+      }, token);
   }
 
   template<typename CompletionToken>
-  auto notify_ack(const Object& o,
-		  const IOContext& ioc,
+  auto notify_ack(Object o,
+		  IOContext ioc,
 		  uint64_t notify_id,
 		  uint64_t cookie,
-		  ceph::buffer::list&& bl,
+		  ceph::buffer::list bl,
 		  CompletionToken&& token) {
-    boost::asio::async_completion<CompletionToken, SimpleOpSig> init(token);
-    notify_ack(o, ioc, notify_id, cookie, std::move(bl),
-	       SimpleOpComp::create(get_executor(),
-				    std::move(init.completion_handler)));
-    return init.result.get();
-  }
-
-  template<typename CompletionToken>
-  auto notify_ack(const Object& o,
-		  std::int64_t pool,
-		  uint64_t notify_id,
-		  uint64_t cookie,
-		  ceph::buffer::list&& bl,
-		  CompletionToken&& token,
-		  std::optional<std::string_view> ns = {},
-		  std::optional<std::string_view> key = {}) {
-    boost::asio::async_completion<CompletionToken, WatchSig> init(token);
-    notify_ack(o, pool, notify_id, cookie, std::move(bl),
-	       SimpleOpComp::create(get_executor(),
-				    std::move(init.completion_handler)),
-	       ns, key);
-    return init.result.get();
+    return boost::asio::async_initiate<CompletionToken, SimpleOpSig>(
+      [o = std::move(o), ioc = std::move(ioc), notify_id, cookie,
+       bl = std::move(bl), this](auto&& handler) mutable {
+	notify_ack(std::move(o), std::move(ioc), std::move(cookie),
+		   std::move(bl), SimpleOpComp::create(get_executor(),
+						       std::move(handler)));
+      }, token);
   }
 
   template<typename CompletionToken>
-  auto unwatch(uint64_t cookie, const IOContext& ioc,
+  auto unwatch(std::uint64_t cookie, IOContext ioc,
```

Comment:
FWIW I don't think sprinkling `std` on trivial integer types is worthwhile.  There is plenty of bare uint64_t uses in this file even after this PR.